src/gazebo_gimbal_controller_plugin.cpp

```diff
@@ -0,0 +1,296 @@
+GZ_REGISTER_MODEL_PLUGIN(GimbalControllerPlugin)
+
+/////////////////////////////////////////////////
+GimbalControllerPlugin::GimbalControllerPlugin()
+  :status("closed")
+{
+  this->pitchPid.Init(1.0, 0, 0, 0, 0, 1.0, -1.0);
+  this->rollPid.Init(1.0, 0, 0, 0, 0, 1.0, -1.0);
+  this->yawPid.Init(1.0, 0, 0, 0, 0, 1.0, -1.0);
+  this->pitchCommand = 0.5* M_PI;  //  is problematic because of singularity
+  this->rollCommand = 0;
+  this->yawCommand = 0;
+}
+
+/////////////////////////////////////////////////
+void GimbalControllerPlugin::Load(physics::ModelPtr _model,
+  sdf::ElementPtr _sdf)
+{
+  this->model = _model;
+
+  this->sdf = _sdf;
+
+  std::string yawJointName = "cgo3_vertical_arm_joint";
+  this->yawJoint = this->model->GetJoint(yawJointName);
+  if (this->sdf->HasElement("joint_yaw"))
+  {
+    // Add names to map
+    yawJointName = sdf->Get<std::string>("joint_yaw");
+    if (this->model->GetJoint(yawJointName))
+    {
+      this->yawJoint = this->model->GetJoint(yawJointName);
+    }
+    else
+    {
+      gzwarn << "joint_yaw [" << yawJointName << "] does not exist?\n";
+    }
+  }
+  if (!this->yawJoint)
+  {
+    gzerr << "GimbalControllerPlugin::Load ERROR! Can't get yaw joint '"
+          << yawJointName << "' " << endl;
+  }
+
+  std::string rollJointName = "cgo3_horizontal_arm_joint";
+  this->rollJoint = this->model->GetJoint(rollJointName);
+  if (this->sdf->HasElement("joint_roll"))
+  {
+    // Add names to map
+    rollJointName = sdf->Get<std::string>("joint_roll");
+    if (this->model->GetJoint(rollJointName))
+    {
+      this->rollJoint = this->model->GetJoint(rollJointName);
+    }
+    else
+    {
+      gzwarn << "joint_roll [" << rollJointName << "] does not exist?\n";
+    }
+  }
+  if (!this->rollJoint)
+  {
+    gzerr << "GimbalControllerPlugin::Load ERROR! Can't get roll joint '"
+          << rollJointName << "' " << endl;
+  }
+
+
+  std::string pitchJointName = "cgo3_camera_joint";
+  this->pitchJoint = this->model->GetJoint(pitchJointName);
+  if (this->sdf->HasElement("joint_pitch"))
+  {
+    // Add names to map
+    pitchJointName = sdf->Get<std::string>("joint_pitch");
+    if (this->model->GetJoint(pitchJointName))
+    {
+      this->pitchJoint = this->model->GetJoint(pitchJointName);
+    }
+    else
+    {
+      gzwarn << "joint_pitch [" << pitchJointName << "] does not exist?\n";
+    }
+  }
+  if (!this->pitchJoint)
+  {
+    gzerr << "GimbalControllerPlugin::Load ERROR! Can't get pitch joint '"
+          << pitchJointName << "' " << endl;
+  }
+
+
+  // get imu sensor
+  std::string imuSensorName = "camera_imu";
+  this->imuSensor = std::static_pointer_cast<sensors::ImuSensor>(
+    sensors::SensorManager::Instance()->GetSensor(imuSensorName));
+  if (this->sdf->HasElement("imu"))
+  {
+    // Add names to map
+    imuSensorName = sdf->Get<std::string>("imu");
+    if (this->model->GetJoint(imuSensorName))
+    {
+      this->imuSensor = std::static_pointer_cast<sensors::ImuSensor>(
```

Comment:
started a new [PR 5295](https://github.com/PX4/Firmware/pull/5295) from my fork, will add the changes in my fork.